src/FSharp.Data.GraphQL.Shared/Validation.fs

```diff
@@ -178,4 +193,199 @@ module Ast =
+            let fragType =
+                match frag.TypeCondition with
+                | Some typeCondition -> 
+                    match schemaInfo.TryGetTypeByName(typeCondition) with
+                    | Some fragType -> fragType
+                    | None -> failwithf "Failure reading schema. Can not find type '%s' specified by a fragment type condition." typeCondition
+                | None -> parentType
+            getSelectionSetInfo schemaInfo fragmentDefinitions fragType frag.SelectionSet
+        selectionSet
+        |> List.collect (function
+            | Field field ->
+                let ifield = parentType.Fields |> Option.map (Array.tryFind (fun f -> f.Name = field.Name)) |> Option.flatten
+                let argDefs = ifield |> Option.map (fun f -> f.Args)
+                let fieldType = ifield |> Option.map (fun f -> f.Type)
+                [ { Alias = field.Alias
+                    Name = field.Name
+                    SelectionSet = getSelectionSetInfo schemaInfo fragmentDefinitions parentType field.SelectionSet
+                    Arguments = field.Arguments
+                    Directives = field.Directives
+                    FieldType = fieldType
+                    ParentType = parentType
+                    ArgDefs = argDefs } ]
+            | InlineFragment frag -> getFragSelectionInfo frag
+            | FragmentSpread spread ->
+                match fragmentDefinitions |> List.tryFind (fun f -> f.Name.IsSome && f.Name.Value = spread.Name) with
+                | Some frag -> getFragSelectionInfo frag
+                | None -> failwithf "Failure reading schema. Can not find fragment '%s'." spread.Name)
+
+    let rec private sameResponseShape (acc : ValidationResult) (fieldA : SelectionInfo, fieldB : SelectionInfo) =
+        if fieldA.FieldType = fieldB.FieldType
+        then
+            let mergedSet = fieldA.SelectionSet |> List.append fieldB.SelectionSet
+            let fieldsForName = mergedSet |> List.groupBy (fun x -> x.AliasOrName)
+            fieldsForName
+            |> List.map snd
+            |> List.fold (fun acc selectionSet ->
+                if selectionSet.Length < 2
+                then acc
+                else List.pairwise selectionSet |> List.fold (fun acc pairs -> sameResponseShape acc pairs) acc) acc
+        else acc @ Error [ sprintf "Field name or alias '%s' appears two times, but they do not have the same return types in the scope of the parent type." fieldA.AliasOrName ]
+
+    let rec private fieldsInSetCanMerge (set : SelectionInfo list) =
+        let fieldsForName = set |> List.groupBy (fun x -> x.AliasOrName)
+        fieldsForName
+        |> List.fold (fun acc (aliasOrName, selectionSet) ->
+            if selectionSet.Length < 2
+            then acc
+            else
+                List.pairwise selectionSet
+                |> List.map (fun pairs -> pairs, sameResponseShape acc pairs)
```

Comment:
You can fuse the 2 maps and the reduce into a singular `List.fold`

Reply:
Done.